### ISSUE TITLE
Occasional loss of cache

### ISSUE THREAD

**Issue by catenocrypt** [catenocrypt/nano-work-cache#7]
Apparently sometimes the cache is lost, reducing in reduced performance and overload situation for the period while the cache is being rebuilt.
Potential reasons (not confirmed):
- Cache modification happens while save is in progress (enumeration not threadsafe)
- If cache file is missing (but backup .bak file is there), nothing is loaded at startup

**Comment by catenocrypt** [comment]
Bak file loading has been added (PR #8 ).

**Comment by catenocrypt** [comment]
Save can crash due to missing map synchronization, verifies using the attached main.go, without the lock + local copy.

```
package main

import (
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)

var housekeepingPeriodSec int = 2

var lastCacheSaveTime int64 = 0

var cacheUpdateTime int64 = 0

var filename = "cachefile"

var workCache map[string]string = map[string]string{}

var workCacheLock = &sync.Mutex{}

func housekeepingCycle() {
	lastCacheSaveTime = cacheUpdateTime
	for {
		doHousekeepingCycle()
		time.Sleep(time.Second * time.Duration(housekeepingPeriodSec))
	}
}

func addingCycle() {
	for {
		doAdding()
		time.Sleep(time.Millisecond * 5)
	}
}

func doAdding() {
	workCacheLock.Lock()
	now := time.Now().Unix()
	n := len(workCache)
	key := fmt.Sprintf("key_%v", n)
	val := fmt.Sprintf("value %v at %v", n, now)
	workCache[key] = val
	workCacheLock.Unlock()
	cacheUpdateTime = now
	//log.Printf("Added %v %v", key, val)
}

func backupFileName(filename string) string {
	return filename + ".bak"
}

func saveCache() {
	// fast exclusive local copy
	startTime := time.Now()
	workCacheLock.Lock()
	n := len(workCache)
	workCacheCopy := make(map[string]string, n)
	for key, entry := range workCache {
		workCacheCopy[key] = entry
	}
	workCacheLock.Unlock()
	elapsed := time.Now().Sub(startTime)
	log.Printf("Save; local copy is made %v %v, dur %v ms", n, len(workCacheCopy), elapsed.Milliseconds())

	// first try to rename old file to .bak (ignore error if not possible / not exists)
	fileNameBak := backupFileName(filename)
	_ = os.Remove(fileNameBak)
	_ = os.Rename(filename, fileNameBak)
	// open new file for writing
	file, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
	if err != nil {
		log.Println("Error saving cache to file, could not open file;", err.Error())
		return
	}
	defer file.Close()

	var cnt int = 0
	for _, entry := range workCacheCopy {
		line := entry
		if len(line) > 0 {
			fmt.Fprintln(file, entry)
			cnt++
		}
	}
	elapsed2 := time.Now().Sub(startTime)
	log.Printf("Cache saved to file, %v entries, dur %v ms", cnt, elapsed2.Milliseconds())
}

func doHousekeepingCycle() {
	//cacheUpdateTime := cacheUpdateTime
	origLastCacheSaveTime := lastCacheSaveTime
	if cacheUpdateTime > lastCacheSaveTime {
		saveCache()
		lastCacheSaveTime = cacheUpdateTime
		log.Printf("Cache saved, %v %v \n", origLastCacheSaveTime, cacheUpdateTime)
	}
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	go housekeepingCycle()
	addingCycle()
}
```

**Comment by catenocrypt** [comment]
Solved by #9 (added locking).
